### ISSUE TITLE
fresh install of numpy picks up older package, cache/hard link issue? - possible conda canary issue

### ISSUE THREAD

**Issue by davidslac** [conda/conda#3260]
I followed a suggestion from Tian to do conda remove -n _test --all, and I think that messed up my numpy. I have started to get an old ImportError, namely ImportError: libopenblasp-r0-39a31c03.2.18.so which I have more detailed notes about in our JIRA system, https://jira.slac.stanford.edu/browse/PSRT-105

To debug this, I have, in two different miniconda installs, done the following

`conda create -n np numpy
`
in the installation that I use regularly, where I am using the conda canary, specifically, or conda info gives

```
np) (psreldev) psel701: /reg/g/psdm/sw/conda $ conda info
Current conda install:

               platform : linux-64
          conda version : 4.2.1
       conda is private : False
      conda-env version : 2.5.2
    conda-build version : 1.21.11+0.g5b44ab3.dirty
         python version : 2.7.12.final.0
       requests version : 2.10.0
       root environment : /reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7  (writable)
    default environment : /reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/envs/np
       envs directories : /reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/envs
          package cache : /reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/pkgs
           channel URLs : file:///reg/g/psdm/sw/conda/channels/system-rhel7
                          file:///reg/g/psdm/sw/conda/channels/psana-rhel7
                          file:///reg/g/psdm/sw/conda/channels/external-rhel7
                          defaults
                          scikit-beam
                          file:///reg/g/psdm/sw/conda/channels/testing-rhel7
            config file : /reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/.condarc
           offline mode : False

```

if I activate the numpy installation and do python -c "import numpy" I get the ImportError on openblas. If in the other installation, where conda info gives

```
(np)(psreldev) psdev105: /reg/g/psdm/sw/conda $ conda info
Current conda install:

             platform : linux-64
        conda version : 4.0.5
  conda-build version : not installed
       python version : 2.7.11.final.0
     requests version : 2.9.1
     root environment : /reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel6  (writable)
  default environment : /reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel6/envs/np
     envs directories : /reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel6/envs
        package cache : /reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel6/pkgs
         channel URLs : file:///reg/g/psdm/sw/conda/channels/system-rhel6/linux-64/
                        file:///reg/g/psdm/sw/conda/channels/system-rhel6/noarch/
                        file:///reg/g/psdm/sw/conda/channels/psana-rhel6/linux-64/
                        file:///reg/g/psdm/sw/conda/channels/psana-rhel6/noarch/
                        file:///reg/g/psdm/sw/conda/channels/external-rhel6/linux-64/
                        file:///reg/g/psdm/sw/conda/channels/external-rhel6/noarch/
                        https://repo.continuum.io/pkgs/free/linux-64/
                        https://repo.continuum.io/pkgs/free/noarch/
                        https://repo.continuum.io/pkgs/pro/linux-64/
                        https://repo.continuum.io/pkgs/pro/noarch/
                        https://conda.anaconda.org/scikit-beam/linux-64/
                        https://conda.anaconda.org/scikit-beam/noarch/
                        file:///reg/g/psdm/sw/conda/channels/testing-rhel7/linux-64/
                        file:///reg/g/psdm/sw/conda/channels/testing-rhel7/noarch/
          config file : /reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel6/.condarc
    is foreign system : False

```

when conda create runs, its output is almost the same between the two, the package plan looks the same:

```
The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    mkl-11.3.3                 |                0       122.1 MB  defaults
    openssl-1.0.2h             |                1         3.2 MB  defaults
    sqlite-3.13.0              |                0         4.0 MB  defaults
    zlib-1.2.8                 |                3         101 KB  defaults
    python-2.7.12              |                1        12.1 MB  defaults
    numpy-1.11.1               |           py27_0         6.2 MB  defaults
    setuptools-25.1.6          |           py27_0         474 KB  defaults
    pip-8.1.2                  |           py27_0         1.5 MB  defaults
    ------------------------------------------------------------
                                           Total:       149.6 MB

The following NEW packages will be INSTALLED:

    mkl:        11.3.3-0      defaults
    numpy:      1.11.1-py27_0 defaults
    openssl:    1.0.2h-1      defaults
    pip:        8.1.2-py27_0  defaults
    python:     2.7.12-1      defaults
    readline:   6.2-2         defaults
    setuptools: 25.1.6-py27_0 defaults
    sqlite:     3.13.0-0      defaults
    tk:         8.5.18-0      defaults
    wheel:      0.29.0-py27_0 defaults
    zlib:       1.2.8-3       defaults

Proceed ([y]/n)? y
```

but what is different is these lines about what gets installed, for the new miniconda where I've never installed a package, I have

```
Fetching packages ...
mkl-11.3.3-0.t 100% |#############################################################################################################################################################################################| Time: 0:00:05  21.35 MB/s
openssl-1.0.2h 100% |#############################################################################################################################################################################################| Time: 0:00:01   2.10 MB/s
sqlite-3.13.0- 100% |#############################################################################################################################################################################################| Time: 0:00:01   2.75 MB/s
zlib-1.2.8-3.t 100% |#############################################################################################################################################################################################| Time: 0:00:00 396.49 kB/s
python-2.7.12- 100% |#############################################################################################################################################################################################| Time: 0:00:02   5.67 MB/s
numpy-1.11.1-p 100% |#############################################################################################################################################################################################| Time: 0:00:01   3.40 MB/s
setuptools-25. 100% |#############################################################################################################################################################################################| Time: 0:00:00 828.33 kB/s
pip-8.1.2-py27 100% |#############################################################################################################################################################################################| Time: 0:00:01   1.34 MB/s
Extracting packages ...
[      COMPLETE      ]|################################################################################################################################################################################################################| 100%
Linking packages ...
[      COMPLETE      ]|################################################################################################################################################################################################################| 100%
```

but for the one where I've been working a while, I get 

```
Linking packages ...
[      COMPLETE      ]|################################################################################################################################################################################################################| 100%
```

everything was already there, and it should be, I only deleted the _test environment, I had other environments with the latest numpy, but they have stopped working.

Now what I really don't understand, is when I compare the files in

`miniconda2/pkgs/numpy-1.11.1-py27_0
`
for the two installations, I see differences. Likewise for the installations in 
`
miniconda2/envs/np/lib/python2.7/site-packages/numpy
`
for instance, in `site-packages/numpy/__init__.py`, I see

```
@@ -108,6 +108,10 @@

 import sys

+import ctypes
+_old_rtld = sys.getdlopenflags()
+sys.setdlopenflags(_old_rtld | ctypes.RTLD_GLOBAL)
+

 class ModuleDeprecationWarning(DeprecationWarning):
     """Module deprecation warning.
@@ -227,3 +231,9 @@
     warnings.filterwarnings("ignore", message="numpy.dtype size changed")
     warnings.filterwarnings("ignore", message="numpy.ufunc size changed")
     warnings.filterwarnings("ignore", message="numpy.ndarray size changed")
+
+sys.setdlopenflags(_old_rtld)
+del _old_rtld
+del sys
+del ctypes
+__mkl_version__ = "11.3.3"
#

```

I don't understand how the same numpy package, with the same name, ends up looking different. This is why I'm submitting it as a conda issue - I think it has to do with cacheing packages and hard links. Maybe the problem strecthes into our filesystem, I can't remember if it is lustre or nfs on this filesystem - this dirty file thing is so strang.

The only thing I can think to do is delete all my environments, run that conda package clean command, and re-install them.

**Comment by HugoTian** [comment]
From your output, it seems that you download numpy in one installation and only linking numpy in the other installation. Could you try to show the output of "conda list numpy" in your 2 installation?

Remove a environment should be safe, it will not break other's functionality. In your numpy-broken environment, could you please try conda update numpy, hope it will fix the error

Thanks

**Comment by davidslac** [comment]
Hi Tian,

I think it looks that way because I had previously downloaded numpy in 
the broken installation. It was working yesterday but stopped working 
today, I know it seems weird, but I think somehow, removing numpy from 
_test rolled back files in the numpy-1.11.1-py_0 package to something 
from 1.10? Maybe that is too crazy.

Here is the output of in the 'np' environment of the two installations, 
I do which python before to verify they are different miniconda 
installations. Note - the broken one has python 2.12, the working one 
python 2.11, but I don't think that should matter.

#######broken:
(np) (psreldev) psel701: /reg/g/psdm/sw/conda $ which python
/reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/envs/np/bin/python
(np) (psreldev) psel701: /reg/g/psdm/sw/conda $ conda list numpy
# packages in environment at 
/reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/envs/np:
#
numpy                     1.11.1                   py27_0


####### working
(np)(psreldev) psdev105: /reg/g/psdm/sw/conda $ which python
/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel6/envs/np/bin/python
(np)(psreldev) psdev105: /reg/g/psdm/sw/conda $ conda list numpy
# packages in environment at 
/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel6/envs/np:
#
numpy                     1.11.1                   py27_0 defaults

**Comment by davidslac** [comment]
the only funny difference above is that the defaults channel wasn't printed for the broken one, but if I add the --show-channel-urls option to conda list, it is defaults.

I tried to update numpy in the broken, but it says it is already up to date.

**Comment by davidslac** [comment]
I removed all environments and references to numpy in the broken miniconda, created a fresh environment with numpy, and it worked. I'm off the rest of the week, but will start building things back up next week, I might take the time to track changes to numpy files between environment updates.

**Comment by davidslac** [comment]
Just one more thought, I ran into something similar that I raised on conda build https://github.com/conda/conda-build/issues/1193. In that case, when I added a test section to the meta.yaml and did a few imports, I was testing an old version of the package -- that is the package files that I was installing were not, or else were not all, making it to the _test environment. I was not uping the build number of the package - so there was an older version to find somewhere.

I think there is some problem with the caching of package files, maybe these issues are related?
